src/messages.js

```diff
@@ -0,0 +1,318 @@
+}
+
+function acceptableContrast(colour, doContrastTest, contrastTestColour, contrastThreshold) {
+  let textContrast = contrast(hexToRgb(colour), hexToRgb(contrastTestColour));
+  return !doContrastTest || textContrast >= contrastThreshold;
+}
+
+function setMessageTextColour(component, settings, member) {
+  let { doContrastTest, contrastTestColour, contrastThreshold } = settings.get();
+  if (member.color && acceptableContrast(member.color, doContrastTest, contrastTestColour, contrastThreshold)) {
+    component.props.style = { color: member.color };
+  }
+}
+
+function handleMessageContent(props, component, settings, profileMap) {
+  const channel = ChannelStore.getChannel(props.message.channel_id);
+
+  if (!channel || !channel.guild_id) return; //No webhooks here lol
+
+  if (settings.get().doColourText) {
+    callbackIfMemberReady(props, profileMap, function (member) {
+      setMessageTextColour(component, settings, member);
+    });
+  }
+}
+
+function patchMessageContent(pluginName, settings, profileMap) {
+  BdApi.Patcher.after(pluginName, MessageContent, 'type', function (_, [props], ret) {
+    handleMessageContent(props, ret, settings, profileMap);
+  });
+}
+
+function handleMessageHeader(props, component, settings, profileMap) {
+  const tree = ZLibrary.Utilities.getNestedProp(component, 'props.username.props.children');
+
+  if (!Array.isArray(tree)) {
+    return;
+  }
+
+  callbackIfMemberReady(props, profileMap, function (member) {
+    let { useServerNames, memberColourPref, tagColourPref, doContrastTest, contrastTestColour, contrastThreshold } =
+      settings.get();
+
+    if (!Object.hasOwn(props.message.author, 'username_real')) {
+      props.message.author.username_real = props.message.author.username.slice();
+
+      if (useServerNames) {
+        // most batshit string length function on earth
+        const count = str => {
+          const regex = /\p{Emoji_Modifier_Base}\p{Emoji_Modifier}?|\p{Emoji_Presentation}|\p{Emoji}\uFE0F|./gu;
+          return ((str || '').match(regex) || []).length;
+        };
+
+        let username_len = count(props.message.author.username_real);
+        let tag_len = count(member.tag);
+
+        props.message.author.username = props.message.author.username_real.slice(0, username_len - tag_len);
+      } else {
+        props.message.author.username = member.name + ' ';
+      }
+    }
+    tree.length = 0; //loser
+
+    let member_tag = member.tag;
+
+    let userProps = {
+      user: props.message.author,
+      className: 'username-h_Y3Us',
+      type: 'member_name',
+    };
+
+    let tagProps = {
+      user: props.message.author,
+      className: 'username-h_Y3Us',
+      type: 'system_tag',
+    };
+
+    // lol
+    let pkBadge = (
+      <span className='botTagCozy-3NTBvK botTag-1NoD0B botTagRegular-kpctgU botTag-7aX5WZ rem-3kT9wc'>
+        <PKBadge
+          pk_id={props.message.id}
+          onClick={id => updateMemberByMsg(id, getUserHash(props.message.author))}
```

Comment:
From some quick testing, the only problem I've discovered is that clicking `PK` badges to refresh a profile throws an error - looks like a missing `profileMap` argument in the call to `updateMemberByMsg` right here.

Reply:
god tell me about it LOL